### PR TITLE
Add deferred loading and caching wrapper for native components

### PR DESCRIPTION
**Pull request by JasonVMo**
### Platforms Impacted
- [x] iOS
- [x] macOS
- [x] win32
- [x] windows
- [x] android

### Description of changes

There are several parts of this change

#### `@fluentui-react-native/native-component` package

This package adds a cached wrapper around `requireNativeComponent`. This addresses two potential gotchas with our previous usage.

**Unnecessary loads**

The naive way to write a component would look something like:

```tsx
const MyNativeComponent = requireNativeComponent('MyNativeComponent');

export const MyComponent = props => {
  const propsInner = doSomeStuff(props);
  return <MyNativeComponent {...propsInner} />;
};
```

The issue here is that the call to `requireNativeComponent` happens when the module is loaded, not when the component is used. This can result in unnecessary overhead as well as causing issues if the native code isn't present. This should also address Issue #228 

**Multiple instances**

For larger projects where different components may be imported directly, it can be easy to end up with slightly different versions of the code present. While this is a performance issue, and can cause subtle behavioral issues if globals are used, in the case of native components this can cause multiple calls to register the same native component. This will cause the view manager to be registered twice resulting in runtime issues.

The package then adds the following:

- `queryNativeComponent` wrapper which provides a safe alternative to `requireNativeComponent`.
- `NativeComponent` and `lateBindComponent` utilities which make it a bit easier to consume the `queryNativeComponent` API

#### Consume the new module

In particular `Callout` and `FocusTrapZone` have switched over to the new paradigm and the frameworks (both old and experimental) support things declared with `NativeComponent`.

#### Move `memo-cache` and `immutable-merge` to 1.0.0

These packages rarely change and feel pretty final in terms of API. Because there is value in keeping the value of the new package consistent it made sense to start it at 1.x and because it depends on memo-cache it made sense to bump that to 1.0 as well.

As part of this I'm renaming `@uifabricshared/immutable-merge` to be `@fluentui-react-native/immutable-merge` as we should be moving away from @uifabricshared as a scope.

### Verification

Automated tests plus manual run through of FocusTrapZone and Callout on Win32 to ensure they still work correctly.

### Pull request checklist

This PR has considered (when applicable):
- [x] Automated Tests
- [x] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/framework/native-component/package.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "@fluentui-react-native/native-component",
+  "version": "1.0.0",
+  "description": "Layered memoization style cache helper",
```
Comment (dotnetjunky) [inline packages/framework/native-component/package.json:4]: Is this description accurate?

Comment (JasonVMo) [reply packages/framework/native-component/package.json]: Good catch.

**Comment by Saadnajmi** [comment]
I imagine I should wait for this to get merged before creating a new Native component for FURN (iOS specifically)?

File: packages/components/Callout/src/Callout.tsx [hunk 2]
```diff
@@ -34,7 +33,7 @@ export const Callout = compose<ICalloutType>({
   },
   settings: settings,
   slots: {
-    root: RCTCallout
+    root: NativeComponent('RCTCallout')
```
Comment (FalseLobster) [inline packages/components/Callout/src/Callout.tsx:36]: It's probably worth checking that the changes you're making here by wrapping the native components with a functional component don't interfere with the setAndForwardRef stuff @PPatBoyd had to do to get useViewCommandFocus working.

Comment (JasonVMo) [reply packages/components/Callout/src/Callout.tsx:36]: In truth they get updated to directly use the host via the `lateBindComponent` call that happens internally. The component wrappers are only point in time wrappers.

**Comment by JasonVMo** [comment]
> I imagine I should wait for this to get merged before creating a new Native component for FURN (iOS specifically)?

Actually I'm probably going to withdraw this after discussing it a bit more with Coates.

**Comment by JasonVMo** [comment]
I'm going to do this a different way, instead looking into a general deferred loading approach for components.

**Pull request closed without merging**